File: Chess/Board.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Chess
{
    public class Board
	{
		private readonly Piece[,] cells;

	    public Board(Piece[,] cells)
	    {
	        this.cells = cells;
	    }

        public IEnumerable<Location> GetPieces(PieceColor color) => 
            AllBoard().Where(loc => GetPiece(loc).Is(color));

        public Piece GetPiece(Location location) => 
            Contains(location) ? cells[location.X, location.Y] : null;

        public void Set(Location location, Piece cell) => 
            cells[location.X, location.Y] = cell;

        public TemporaryPieceMove PerformTemporaryMove(Location from, Location to)
		{
			var old = GetPiece(to);
			Set(to, GetPiece(from));
			Set(from, null);
			return new TemporaryPieceMove(this, from, to, old);
		}

        private IEnumerable<Location> AllBoard()
        {
            return 
                from y in Enumerable.Range(0, cells.GetLength(0))
                from x in Enumerable.Range(0, cells.GetLength(1))
                select new Location(x, y);
        }

        public bool Contains(Location loc) =>
            loc.X >= 0 && loc.X < cells.GetLength(0) && 
            loc.Y >= 0 && loc.Y < cells.GetLength(1);
	}

	public class TemporaryPieceMove : IDisposable
	{
		private readonly Board board;
		private readonly Location from;
		private readonly Piece oldDestinationPiece;
		private readonly Location to;

		public TemporaryPieceMove(Board board, Location from, Location to, Piece oldDestinationPiece)
		{
			this.board = board;
			this.from = from;
			this.to = to;
			this.oldDestinationPiece = oldDestinationPiece;
		}

		public void Undo()
		{
			board.Set(from, board.GetPiece(to));
			board.Set(to, oldDestinationPiece);
		}

	    public void Dispose()
	    {
	        Undo();
	    }
	}
}

File: python/Chess/Solved/board.py
from __future__ import annotations
from typing import List, Optional
from location import Location
from temporary_piece_move import TemporaryPieceMove
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from piece_color import PieceColor
    from piece import Piece


class Board:
    def __init__(self, cells: List[List[Optional[Piece]]]):
        self.cells = cells

    def get_pieces(self, color: PieceColor):
        return [
            location
            for location in self.all_board()
            if self.get_piece(location) and self.get_piece(location).color == color
        ]

    def get_piece(self, location: Location):
        return self.cells[location.y][location.x] if self.contains(location) else None

    def set(self, location: Location, cell: Optional[Piece]):
        self.cells[location.y][location.x] = cell

    def perform_temporary_move(self, location_from: Location, location_to: Location):
        old = self.get_piece(location_to)
        self.set(location_to, self.get_piece(location_from))
        self.set(location_from, None)
        return TemporaryPieceMove(self, location_from, location_to, old)

    def all_board(self):
        result = []

        for y in range(len(self.cells)):
            for x in range(len(self.cells[0])):
                result.append(Location(x, y))

        return result

    def contains(self, location: Location):
        return 0 <= location.x < len(self.cells[0]) and 0 <= location.y < len(
            self.cells
        )
